lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,179 @@
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+          opts.on('--all', 'Operate on all certnames') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        if input['all']
+          requested_certnames = get_all_pending_certs(puppet.settings)
+          if requested_certnames.nil?
+            return 1
+          end
+        else
+          requested_certnames = input['certname']
+        end
+
+        success = sign_requested_certs(requested_certnames, puppet.settings)
+        return success ? 0 : 1
+      end
+
+      def get_certificate_statuses(settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_statuses', 'any_key')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def sign_cert(certnames,settings)
+        results = { }
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_status')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          certnames.each do |certname|
+            url.resource_name = certname
+            results[certname] = connection.put(BODY, url)
+          end
+        end
+        return results
+      end
+
+      def get_all_certs(settings)
+        result = get_certificate_statuses(settings)
+
+        unless result.is_a? Net::HTTPOK
+            @logger.err 'Error:'
+            @logger.err "#{result.inspect}"
+            return nil
+        end
+        return result
+      end
+
+      def select_pending_certs(get_result)
+        requested_certnames = JSON.parse(get_result).select{|e| e["state"] == "requested"}.map{|e| e["name"]}
+
+        if requested_certnames.empty?
+          @logger.err 'Error'
+          @logger.err "No waiting certificate requests to sign"
+          return nil
+        end
+
+        return requested_certnames
+      end
+
+      def get_all_pending_certs(settings)
+        result = get_all_certs(settings)
+        if result
+          select_pending_certs(result.body)
+        end
+      end
+
+      def sign_requested_certs(certnames,settings)
+        success = true
+        results = sign_cert(certnames, settings)
+        results.each do |certname, result|
+          case result.code
+          when '204'
```

Comment:
are we sure it returns a 204 and not a 200?  I think 204 is standard...but

Reply:
I'm sure. We get 204 back because it's not returning a body, but the request was successful. There are some comments in the PR somewhere (collapsed now because of a lot of squashing) where I talked about this with @Magisus

Reply:
Thanks!  I just noticed that submitting the csr to CA returns a 200 even though it doesnt return content.  So wanted to double check with yours